File: day11/day11.py
def transform_stone(stone: int) -> list[int]:
    digits = str(stone)
    length = len(digits)
    if stone == 0:
        return [1]
    elif length % 2 == 0:
        return [int(digits[: length // 2]), int(digits[length // 2 :])]
    else:
        return [stone * 2024]


memo: dict[tuple[int, int], int] = {}


def count_stones(n: int, stone: int) -> int:
    if (n, stone) in memo:
        return memo[(n, stone)]
    elif n == 1:
        return len(transform_stone(stone))
    next_stones = transform_stone(stone)
    count = 0
    for next_stone in next_stones:
        length = count_stones(n - 1, next_stone)
        memo[(n - 1, next_stone)] = length
        count += length
    return count


def count_all(n: int, stones: list[int]) -> int:
    count = 0
    for stone in stones:
        count += count_stones(n, stone)
    return count


stones = [int(stone) for stone in open("input").read().split()]

print(count_all(25, stones))
print(count_all(75, stones))

File: day11/Day11.hs
import Data.IntMultiSet (IntMultiSet)
import qualified Data.IntMultiSet as MS

parse :: String -> [Int]
parse = map read . words

transformOne :: Int -> [Int]
transformOne 0 = [1]
transformOne stone
  | evenDigits = [read left, read right]
  | otherwise = [stone * 2024]
  where
    digits = show stone
    len = length digits
    evenDigits = len `mod` 2 == 0
    (left, right) = splitAt (len `div` 2) digits

transformAll :: IntMultiSet -> IntMultiSet
transformAll = MS.concatMap transformOne

solve :: Int -> [Int] -> Int
solve n = MS.size . (!! n) . iterate transformAll . MS.fromList

part1 :: [Int] -> Int
part1 = solve 25

part2 :: [Int] -> Int
part2 = solve 75

main = do
  contents <- readFile "input"
  let stones = parse contents
  print $ part1 stones
  print $ part2 stones
